source/common/router/scoped_rds.cc

```diff
@@ -98,77 +99,109 @@ void ScopedRdsConfigSubscription::onConfigUpdate(
+        throw EnvoyException(
+            fmt::format("scope key conflict found, first scope is '{}', second scope is '{}'",
+                        iter->second, scoped_route_info->scopeName()));
+      }
+      scope_name_by_hash_[scoped_route_info->scopeKey().hash()] = scoped_route_info->scopeName();
+      scoped_route_map_[scoped_route_info->scopeName()] = scoped_route_info;
+      applyDeltaConfigUpdate(
+          [scoped_route_info](const ConfigProvider::ConfigConstSharedPtr& config) {
+            auto* thread_local_scoped_config = const_cast<ThreadLocalScopedConfigImpl*>(
+                static_cast<const ThreadLocalScopedConfigImpl*>(config.get()));
 
-      thread_local_scoped_config->addOrUpdateRoutingScope(scoped_route_info);
-    });
+            thread_local_scoped_config->addOrUpdateRoutingScope(scoped_route_info);
+          });
+      any_applied = true;
+      ENVOY_LOG(debug, "srds: add/update scoped_route '{}'", scoped_route_info->scopeName());
+    } catch (const EnvoyException& e) {
+      exception_msgs.emplace_back(fmt::format("{}", e.what()));
+    }
   }
-  for (const auto& scoped_route_name : removed_resources) {
-    ENVOY_LOG(debug, "srds: remove scoped route '{}'", scoped_route_name);
-    scoped_route_map_.erase(scoped_route_name);
-    applyDeltaConfigUpdate([scoped_route_name](const ConfigProvider::ConfigConstSharedPtr& config) {
-      auto* thread_local_scoped_config = const_cast<ThreadLocalScopedConfigImpl*>(
-          static_cast<const ThreadLocalScopedConfigImpl*>(config.get()));
-      thread_local_scoped_config->removeRoutingScope(scoped_route_name);
-    });
+  for (const auto& scope_name : removed_resources) {
+    auto iter = scoped_route_map_.find(scope_name);
+    if (iter != scoped_route_map_.end()) {
+      scope_name_by_hash_.erase(iter->second->scopeKey().hash());
+      scoped_route_map_.erase(iter);
+      applyDeltaConfigUpdate([scope_name](const ConfigProvider::ConfigConstSharedPtr& config) {
+        auto* thread_local_scoped_config = const_cast<ThreadLocalScopedConfigImpl*>(
+            static_cast<const ThreadLocalScopedConfigImpl*>(config.get()));
+        thread_local_scoped_config->removeRoutingScope(scope_name);
+      });
+      any_applied = true;
+      ENVOY_LOG(debug, "srds: remove scoped route '{}'", scope_name);
+    }
   }
   ConfigSubscriptionCommonBase::onConfigUpdate();
-  setLastConfigInfo(absl::optional<LastConfigInfo>({absl::nullopt, version_info}));
+  if (any_applied) {
+    setLastConfigInfo(absl::optional<LastConfigInfo>({absl::nullopt, version_info}));
+  }
   stats_.config_reload_.inc();
+  if (!exception_msgs.empty()) {
+    throw EnvoyException(fmt::format("Error adding/updating scoped route(s): {}",
+                                     StringUtil::join(exception_msgs, ", ")));
+  }
 }
 
+// TODO(stevenzzzz): consider generalizing this function as it overlaps with
+// CdsApiImpl::onConfigUpdate.
 void ScopedRdsConfigSubscription::onConfigUpdate(
```

Comment:
Please add TODO referencing config rejection discrepancy with delta `onConfigUpdate()` and issue #7508.

Reply:
done

Reply:
done